lightning/src/ln/channelmanager.rs

```diff
@@ -4715,13 +4746,35 @@ where
 	/// event matches your expectation. If you fail to do so and call this method, you may provide
 	/// the sender "proof-of-payment" when they did not fulfill the full expected payment.
 	///
+	/// This function will fail the payment if it has custom TLVs with even type numbers, as we
+	/// will assume they are unknown. If you intend to accept even custom TLVs, you should use
+	/// [`claim_funds_with_known_custom_tlvs`].
+	///
 	/// [`Event::PaymentClaimable`]: crate::events::Event::PaymentClaimable
 	/// [`Event::PaymentClaimable::claim_deadline`]: crate::events::Event::PaymentClaimable::claim_deadline
 	/// [`Event::PaymentClaimed`]: crate::events::Event::PaymentClaimed
 	/// [`process_pending_events`]: EventsProvider::process_pending_events
 	/// [`create_inbound_payment`]: Self::create_inbound_payment
 	/// [`create_inbound_payment_for_hash`]: Self::create_inbound_payment_for_hash
+	/// [`claim_funds_with_known_custom_tlvs`]: Self::claim_funds_with_known_custom_tlvs
 	pub fn claim_funds(&self, payment_preimage: PaymentPreimage) {
+		self.claim_payment_internal(payment_preimage, false);
+	}
+
+	/// This is a variant of [`claim_funds`] that allows accepting a payment with custom TLVs with
+	/// even type numbers.
+	///
+	/// # Note
+	///
+	/// You MUST check you've understood all even TLVs before using this to
+	/// claim, otherwise you may unintentionally agree to some protocol you do not understand.
```

Comment:
I don’t know if it should be documented here, in the pending changelog about this new custom HTLC TLVs support or at the spec-level, though as a receiver it should be mandatory to have a receiver-issued secret as part of the custom TLV before to even accept to process an experimental `tlv_record` from an _unknown_ LN sender.

See the concern above deanonymization attack, and how such requirement would serve as a deterrent.

Reply:
Let’s take the “Cancellable payments” protocol as described here: https://github.com/BlockstreamResearch/scriptless-scripts/blob/master/md/multi-hop-locks.md#cancellable-payments

Let’s assume there is a `valid_until_chain_height` negotiated as part of the protocol after which the multi-shard payments should be all cancelled by the recipient back to the sender.

Lack of authentication by the mean of a receiver-issued secret allow anyone with the details of the cancellable payments to send a payment shard (e.g a routing hop), and therefore to probe the chain height as seen by the recipient Lightning node (a fingerprint vector that could be used to cross-link with the on-chain full-node).

The landscape could be different with PTLCs, though here it’s too early to say if/how mutli-shard payment could be carried through. Yes, PTLCs could come with an inherent access structure, though this is not certain as of today.

Overall, it’s a reasonable internet protocol design rule to authenticate your network input as early as you can. Recommending our custom HTLC TLVs to authenticate its custom protocol sounds a good rule of thumb.

For recent failure between Lightning API interface and application processing, see https://lists.linuxfoundation.org/pipermail/lightning-dev/2023-June/003983.html

Reply:
> can you describe an attack where payment_secret isn't sufficient and something has to be added to the custom TLVs?

Yes where you have multiple payers and one payee, and each have to pay their shares of the payment e.g a crowdfunding between named participants. So current `payment_secret` wouldn’t work and you would need a specific authentication flow.

Somehow that type of multi-payer single-payee for crowdfunding or common bill splitting is what custom HTLC TLVs would be very nice to experiment with.

Reply:
Hmm, I'm not sure doing authentication with custom TLVs within a multi-payer single-payee MPP payment is possible in the current implementation of custom TLVs as we drop/fail non-matching TLVs, so if different payment parts had different secrets, they'd just get dropped?

Although, either way, I think for now this recommendation may be a little out of scope for the docs here

Reply:
Well you would need to have custom TLVs for `multipayer_secret` where the value is different for each payment parts, and application-level logic to account the TLV. Though I’ve not looked if it’s feasible and working well with current event model.

The hard thing is it’s hard to see if authentication or privacy is screwed or not before you have both a specification of the protocol and implementation, I think the sentence “You MAY authenticate the even TLVs additionally of `payment_secret` to match your protocol semantic” might be reasonable.

Reply:
Sorry was late to respond to this, is it helpful to add this even though a user couldn't actually take the suggestion? Ultimately any different custom TLV values between different payment parts will be dropped/failed through `RecipientOnionFields::check_merge` in our MPP validation, so the user wouldn't be able to authenticate even TLVs separately from payment secret. If in the future our custom TLV logic were changed to provide the user with individual payment parts' custom TLVs, then this seems fair, but as the code currently stands I'm not sure it makes sense to include?

Reply:
> Ultimately any different custom TLV values between different payment parts will be dropped/failed through RecipientOnionFields::check_merge in our MPP validation, so the user wouldn't be able to authenticate even TLVs separately from payment secret. 

Fair point, multi-payer secret would need change in our MPP validation, well kinda restraint what you can do in term of custom HTLC TLVs, though can be extended in the future.